Update the readme file (new template) (#1)
DevExpress-Examples/wpf-grid-sync-isnodeexpanded-with-view-model: Readme.md

[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
[![](https://img.shields.io/badge/💬_Leave_Feedback-feecdd?style=flat-square)](#does-this-example-address-your-development-requirementsobjectives)
<!-- default badges end -->

# WPF Grid (TreeListView) - Sync TreeListNode Expansion with ViewModel

> This example adds two-way synchronization between the expansion state of [`TreeListNode`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode) and a `Boolean` property in the `ViewModel`.

Reviewer: The first question I'd like to ask.

> This example **adds** two-way synchronization ...

If you say "add", does it mean that previously there was some other synchronization? Or does this example demonstrate two-way synchronization as the only synchronization method from scratch?

Second. The **expansion** state.

Well, "expansion" is not a correct term in this scenario because the first association with "expansion" is "territorial expansion" and similar.

If we speak of a node state, this is "the expanded node state" or "the expanded state".

Third.

> state of `TreeListNode`

Do we speak of a single tree list node (as an instance) or we speak of all tree list nodes in this example/application?

This part suggests that we just map one node to another some Boolean property:

> synchronization between the expanded state of `TreeListNode` and a `Boolean` property...

We need to clarify this introduction. Also, in the perfect world, I would like to see more clarity -- why this scenario is important/valuable (in terms of practical sense)? This example implements such a synchronization? OK. But what's the real-world scenario lies at the core of our technical solution? As a clear example.

Author: This example implements two-way synchronization between the expanded state of each [`TreeListNode`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode) and a `Boolean` property in the bound data object.

Reviewer: OK. I see...

```suggestion
This example adds two-way synchronization between the expanded state of each [`TreeListNode`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode) and a `Boolean` property in the `ViewModel`.
```

![Sync TreeListNode Expansion with ViewModel](./Images/grid.jpg)

> DevExpress does not expose the [`IsExpanded`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode.IsExpanded) property as a `DependencyProperty`, so direct data binding is not possible. This example uses a custom behavior (`BindableExpandingBehavior`) to simulate two-way binding and keep the [`TreeListView`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListView) state in sync with the underlying `ViewModel`.

Reviewer: > DevExpress does not expose ...

This phrase sounds weird to me...  Besides... should it be here..? Maybe we should say something simpler. For example:

> `TreeListNode.IsExpanded` is not a dependency property, therefore, direct data binding is unavailable.

OK. Then.

> This example uses a custom behavior (`BindableExpandingBehavior`) ...

What is "custom behavior" in this context? Is it an API entity implemented as part of this example? `BindableExpandingBehavior` -- is it a type/class name?

>  to simulate two-way binding and keep the [`TreeListView`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListView) state in sync with the underlying `ViewModel`.

There is a better way of expressing this idea but not now. First, we need more clarity regarding this scenario.

Author: [`TreeListNode.IsExpanded`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode.IsExpanded) is not a dependency property. You cannot bind it directly to a `Boolean` property in your data model.

This example defines a helper class (`BindableExpandingBehavior`) that attaches to a `GridControl` and synchronizes the expanded state of nodes with the `IsExpanded` property in bound data objects.

The `BindableExpandingBehavior` class updates each node when the underlying property changes. It also monitors expand/collapse actions in the UI and updates the corresponding data property to reflect the new state.

Reviewer: ```suggestion
[`IsExpanded`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode.IsExpanded) is not a dependency property. Therefore, you cannot bind this property directly to a `Boolean` property in your data model.

This example defines a helper class (`BindableExpandingBehavior`) that attaches to a `GridControl` and synchronizes the expanded state for each node with the [`IsExpanded`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode.IsExpanded) property in bound data objects.

The `BindableExpandingBehavior` class updates each node in response to underlying property value changes. This helper class also tracks expand/collapse actions in the UI and updates the corresponding data property to reflect the change.
```

> Use this technique to:

Reviewer: ```suggestion
You can use this technique to:
```

> * Keep expansion states consistent across user sessions.

Reviewer: Expansion...

```suggestion
* Maintain node expanded state consistency across user sessions.
```

> * Programmatically control expanded/collapsed states from the data model.

Reviewer: ```suggestion
* Manage node expanded/collapsed states in code at the data model level.
```

> * Respond to UI actions through the `ViewModel`.

Reviewer: ```suggestion
* Respond to UI actions using the `ViewModel`.
```

> You can reuse the `BindableExpandingBehavior` in any [`GridControl`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.GridCell.GridControl) that uses a hierarchical `TreeListView`.

Reviewer: > You can reuse the `BindableExpandingBehavior` 

Again, what is it? OK. We see below, it's a class.

reuse -- uses...

```suggestion
You can reuse the `BindableExpandingBehavior` class in any [`GridControl`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.GridCell.GridControl) that includes a hierarchical `TreeListView`.
```

## Implementation Details

### Bindable Expanding Behavior

> The `BindableExpandingBehavior` class syncs `TreeListNode` expansion with a `Boolean` property in your `ViewModel`:

Reviewer: ```suggestion
The `BindableExpandingBehavior` class synchronizes the `TreeListNode` expanded state with a `Boolean` property in your `ViewModel`:
```

> * Attach the behavior to a `GridControl` and set the `ExpandingProperty` to the name of your `Boolean` property (for example, `IsExpanded`).

Reviewer: OK. Who does the written actions exactly?

> Attach the behavior

is in imperative mood, as if it's addressed to the reader directly rather than a description of what the example does... Moreover, this section is called "Implementation Details" after all...

Author: The example attaches a `BindableExpandingBehavior` to the `GridControl` and sets its `ExpandingProperty` to the name of a `Boolean` property (`IsExpanded`) declared in each data item.

Reviewer: ```suggestion
* The example attaches `BindableExpandingBehavior` to the `GridControl` and sets its `ExpandingProperty` to `IsExpanded` (the name of the property declared in each data item).
```

> * When the `IsExpanded` property changes in the data object, the corresponding node in the tree expands or collapses automatically.

Reviewer: OK. Suppose, these items are ought to describe what the example/demo does...

"Property changes" can be read as:

1. Property name changes
2. Property type changes
3. Property value changes

```suggestion
* Every time the `IsExpanded` property value changes in the data object, the corresponding node expands or collapses automatically.
```

> * When a user expands or collapses a node in the UI, the behavior updates the `IsExpanded` property in the bound object.

Reviewer: Does a user have an ability to expand or collapse a node outside UI in this scenario?

> the behavior updates

"the behavior" looks strange (as a term)

Author: When I write 'the behavior', I mean the BindableExpandingBehavior class. So the sentence will be as follows: When a user expands or collapses a node in the UI, the `BindableExpandingBehavior` class updates the `IsExpanded` property in the bound object.

Reviewer: OK, I see.

```suggestion
* When a user expands or collapses a node in the UI, `BindableExpandingBehavior` updates the `IsExpanded` property value for the bound object.
```

> The logic is handled in two directions:

Reviewer: > The **logic** is handled in two directions:

Normally, I object against using the word "logic". Indeed, this word has this meaning in Russian, but this fact does not allow you to blindly map the same idea in English.

Besides, this phrase is just worded vaguely on its own...

```suggestion
State synchronization is implemented as two event handlers corresponding to forward and reverse directions:
```

> 1. When data changes

Reviewer: ```suggestion
1. Forward synchronization (when data changes)
```

```csharp
public void PropertyChanged(object obj, PropertyChangedEventArgs args) {
    if(args.PropertyName == this.ExpandingProperty) {
        var node = FindNodeByValue(obj, treeView.Nodes);
        node.IsExpanded = (bool)obj.GetType().GetProperty(args.PropertyName).GetValue(obj);
    }
}
```

> 2. When the user expands/collapses a node

Reviewer: ```suggestion
2. Reverse synchronization (when a user expands/collapses a node)
```

```csharp
void GridNodeChanged(object sender, TreeListNodeEventArgs e) {
    var propInfo = e.Node.Content.GetType().GetProperty(this.ExpandingProperty);
    propInfo.SetValue(e.Node.Content, e.Node.IsExpanded);
}
```

### Data Model

> `Parent` and `Child` classes both expose the `IsExpanded` property that participates in synchronization:

Reviewer: ```suggestion
Both `Parent` and `Child` classes expose the `IsExpanded` property required for synchronization:
```

```csharp
public bool IsExpanded {
    get { return this._IsExpanded; }
    set { this.SetProperty(ref this._IsExpanded, value, "IsExpanded"); }
}
```

> Each `Parent` contains a list of `Child` objects, and each `Child` can contain a list of `Toy` objects, creating a three-level tree structure.

Reviewer: ```suggestion
Each `Parent` contains a list of `Child` objects. In turn, each `Child` can contain a list of `Toy` objects, therefore, creating a three-level tree structure.
```

> The `DataHelper` class generates a large collection of `Parent` objects with nested children.

Reviewer: ```suggestion
The `DataHelper` class generates a large collection of `Parent` objects populated with nested `Child` objects.
```

### Setup View

The main window binds the `GridControl` to an `ObservableCollection<Parent>`. Each `Parent` has a collection of `Child` objects, and each `Child` has a collection of `Toy` items. The `TreeListView` displays this hierarchy across three levels.

> To synchronize node expansion with the `ViewModel`, the grid includes the `BindableExpandingBehavior`:

Reviewer: ```suggestion
To synchronize node expanded states with the `ViewModel`, the grid includes the `BindableExpandingBehavior` class as follows:
```

[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
[![](https://img.shields.io/badge/💬_Leave_Feedback-feecdd?style=flat-square)](#does-this-example-address-your-development-requirementsobjectives)
<!-- default badges end -->

# WPF Grid (TreeListView) - Sync TreeListNode Expansion with ViewModel

This example adds two-way synchronization between the expanded state of each [`TreeListNode`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode) and a `Boolean` property in the `ViewModel`.

![Sync TreeListNode Expansion with ViewModel](./Images/grid.jpg)

> [`IsExpanded`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode.IsExpanded) is not a dependency property. Therefore, you cannot bind this property directly to a `Boolean` property in your data model.

Reviewer 2: ```suggestion
[`IsExpanded`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode.IsExpanded) is not a dependency property - you cannot bind this property directly to a `Boolean` property in your data model.
```

This example defines a helper class (`BindableExpandingBehavior`) that attaches to a `GridControl` and synchronizes the expanded state for each node with the [`IsExpanded`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.TreeListNode.IsExpanded) property in bound data objects.

The `BindableExpandingBehavior` class updates each node in response to underlying property value changes. This helper class also tracks expand/collapse actions in the UI and updates the corresponding data property to reflect the change.

You can use this technique to:

* Maintain node expanded state consistency across user sessions.

* Manage node expanded/collapsed states in code at the data model level.

* Respond to UI actions using the `ViewModel`.

You can reuse the `BindableExpandingBehavior` class in any [`GridControl`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Grid.GridCell.GridControl) that includes a hierarchical `TreeListView`.

## Implementation Details

### Bindable Expanding Behavior

> The `BindableExpandingBehavior` class synchronizes the `TreeListNode` expanded state with a `Boolean` property in your `ViewModel`:

Reviewer 2: ```suggestion
The `BindableExpandingBehavior` class synchronizes the `TreeListNode`'s expanded state with a `Boolean` property in your `ViewModel`:
```

* When a user expands or collapses a node in the UI, `BindableExpandingBehavior` updates the `IsExpanded` property value for the bound object.

State synchronization is implemented as two event handlers corresponding to forward and reverse directions:

1. Forward synchronization (when data changes)

```csharp
public void PropertyChanged(object obj, PropertyChangedEventArgs args) {
    if(args.PropertyName == this.ExpandingProperty) {
        var node = FindNodeByValue(obj, treeView.Nodes);
        node.IsExpanded = (bool)obj.GetType().GetProperty(args.PropertyName).GetValue(obj);
    }
}
```

2. Reverse synchronization (when a user expands/collapses a node)

```csharp
void GridNodeChanged(object sender, TreeListNodeEventArgs e) {
    var propInfo = e.Node.Content.GetType().GetProperty(this.ExpandingProperty);
    propInfo.SetValue(e.Node.Content, e.Node.IsExpanded);
}
```

### Data Model

Both `Parent` and `Child` classes expose the `IsExpanded` property required for synchronization:

```csharp
public bool IsExpanded {
    get { return this._IsExpanded; }
    set { this.SetProperty(ref this._IsExpanded, value, "IsExpanded"); }
}
```

> Each `Parent` contains a list of `Child` objects. In turn, each `Child` can contain a list of `Toy` objects, therefore, creating a three-level tree structure.

Reviewer 2: ```suggestion
Each `Parent` contains a list of `Child` objects. In turn, each `Child` can contain a list of `Toy` objects, creating a three-level tree structure.
```
